torch/distributed/_tools/mem_tracker.py

```diff
@@ -0,0 +1,920 @@
+            continue
+        print(
+            f"Device: {dev}",
+            *(
+                f"\t{k}: {_rounding_fn(v, divisor, 2)} {units}"
+                for k, v in dev_snap.items()
+            ),
+            sep="\n",
+        )
+
+
+def _print_snapshot_tabular(
+    snapshot: Dict[torch.device, Dict[str, int]], units: str
+) -> None:
+    if len(snapshot) == 0:
+        print("No memory tracked.")
+        return
+    try:
+        from tabulate import tabulate
+
+        divisor = _get_mem_divisor(units)
+        table_data = []
+        key_list = list(next(iter(snapshot.values())).keys())
+        headers = ["Device"] + [f"{key}" for key in key_list]
+
+        for dev, dev_snap in snapshot.items():
+            if _rounding_fn(dev_snap[_TOTAL_KEY], divisor, 2) <= 0:
+                continue
+            row = [str(dev)]
+            row.extend(
+                f"{_rounding_fn(v, divisor, 2)} {units}" for v in dev_snap.values()
+            )
+            table_data.append(row)
+
+        print(tabulate(table_data, headers=headers, tablefmt="rst"))
+    except ImportError as err:
+        raise ImportError(
+            "Please install tabulate to use the tabulate option."
+        ) from err
+
+
+def _print_state_snapshots(
+    snapshots: Dict[_State, List[Dict[torch.device, Dict[str, int]]]], units: str
+) -> None:
+    for state, snapshot_list in snapshots.items():
+        print(f"{state}")
+        for i, snapshot in enumerate(snapshot_list):
+            print(f"# {i + 1}:")
+            _print_snapshot(snapshot, units)
+    print()
+
+
+def _print_state_snapshots_tabular(
+    snapshots: Dict[_State, List[Dict[torch.device, Dict[str, int]]]], units: str
+) -> None:
+    try:
+        from tabulate import tabulate
+
+        table_data = []
+        last_state_call = None
+        divisor = _get_mem_divisor(units)
+        for state, snapshot_list in snapshots.items():
+            for i, snapshot in enumerate(snapshot_list):
+                state_call = f"{state} # {i + 1}"
+                for dev, dev_snap in snapshot.items():
+                    if _rounding_fn(dev_snap[_TOTAL_KEY], divisor, 2) <= 0:
+                        continue
+                    row = {
+                        "State & Call": (
+                            state_call if state_call != last_state_call else ""
+                        ),
+                        "Device": str(dev),
+                    }
+                    last_state_call = state_call
+                    for k, v in dev_snap.items():
+                        row[f"{k}"] = f"{_rounding_fn(v, divisor, 2)} {units}"
+                    table_data.append(row)
+        print(tabulate(table_data, headers="keys", tablefmt="rst"))
+
+    except ImportError as err:
+        raise ImportError(
+            "Please install tabulate to use the tabulate option."
+        ) from err
+
+
+class _UpdateType(Enum):
+    ADD = auto()
+    DEL = auto()
+    REF = auto()
```

Comment:
I guess `REF` and `SIZE` are not obvious to me and require looking more closely at the impl.

Reply:
Will add some comments.